examples/provider_upgrade_test.go

```diff
@@ -0,0 +1,411 @@
+// Copyright 2016-2023, Pulumi Corporation.  All rights reserved.
+
+//go:build java || all
+// +build java all
+
+// Experimental provider upgrade tests.
+//
+// The objective of these tests is to make sure the provider release candidate will not generate any
+// Update or Replace plans for Pulumi stacks deployed on a previous baseline version of the
+// provider.
+//
+// Note on build flags above --^ these tests currently uses YAML only but our build workflow matrix
+// does not run YAML-only test a the moment, and it splits tests by language build tag; for this
+// reason the file is marked as another language so that the test runs only once.
+
+package examples
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	jsonpb "google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"gopkg.in/yaml.v2"
+
+	pfbridge "github.com/pulumi/pulumi-terraform-bridge/pf/tfbridge"
+	testutils "github.com/pulumi/pulumi-terraform-bridge/testing/x"
+	"github.com/pulumi/pulumi/pkg/v3/testing/integration"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+	pulumirpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
+
+	providerRC "github.com/pulumi/pulumi-aws/provider/v6"
+	"github.com/pulumi/pulumi-aws/provider/v6/pkg/version"
+)
+
+const (
+	acceptEnvVar   = "PULUMI_ACCEPT"
+	providerBinary = "pulumi-resource-aws"
+)
+
+var (
+	accept bool = cmdutil.IsTruthy(os.Getenv(acceptEnvVar))
+)
+
+func TestProviderUpgradeQuick(t *testing.T) {
+	info := newProviderUpgradeInfo(t)
+
+	bytes, err := os.ReadFile(info.grpcFile)
+	require.NoErrorf(t, err, "No pre-recorded gRPC log found, try to run TestProviderUpgradeRecord")
+
+	eng := &mockPulumiEngine{
+		provider:              providerServer(t),
+		lastCheckRequestByURN: map[string]*pulumirpc.CheckRequest{},
+	}
+
+	for _, line := range strings.Split(string(bytes), "\n") {
+		line = strings.TrimSpace(line)
+		if line == "" {
+			continue
+		}
+		line = ignoreStables(t, line)
+		eng.replayGRPCLog(t, line)
+	}
+	require.NotEmptyf(t, eng.verifiedDiffResourceCounter, "Need at least one replay test")
+}
+
+func TestProviderUpgradeRecord(t *testing.T) {
+	if !accept {
+		t.Skipf("Skipping; to record baselines set %s env var to true", acceptEnvVar)
+	}
```

Comment:
We normally have this in a single test, that is always run in either a "record" or "replay" mode. We might want to do that here, since `TestProviderUpgradeQuick` and `TestProviderUpgradeRecord` interact via the file system.

Since it never makes sense to run both tests in 1 go, we might want to merge them here as well.

Reply:
Edited for it to sit in a single test. I do think refreshing belongs in the test world and not a separate binary, more idiomatic  Go it seems, see https://github.com/hexops/autogold again.